### PR TITLE
Feat/304 notification repo local

### LINKED ISSUE #304: Notification repository local
Local variant of the Notifications repository, following the established and merged interface. (#292)

### PR DESCRIPTION
**Pull request by AACioffi**
## What?
Implements the in-memory local repository for notifications with comprehensive tests. Closes #304.

## Why?
The notification system requires:
- **A local repository for ViewModel unit tests** - VMs should be tested against a lightweight in-memory implementation (fast, deterministic), not Firebase
- A reference implementation demonstrating the NotificationsRepository interface

## How?
**Implementation (`NotificationsLocalRepository.kt`):**

- Implemented all `NotificationsRepository` interface methods using in-memory `MutableMap<String, Notification>` storage:
  - `getNewId()`: Counter-based ID generation (increments from 0)
  - `addNotification()`: Stores notification by ID as map key
  - `getNotification()`: Retrieves by ID, throws `NoSuchElementException` if not found
  - `getUserNotifications()`: Filters by `recipientId` and returns sorted list (descending by `emissionTime`)
  - `markAsRead()`: Creates updated copy with `wasRead = true`
  - `deleteNotification()`: Removes from map, throws `NoSuchElementException` if not found

**Test Suite (`NotificationsLocalRepositoryTest.kt`):**

- 12 comprehensive tests organized by category:
  - **Happy paths** (5 tests): ID uniqueness, add/retrieve, user filtering, mark as read, delete
  - **Edge cases** (4 tests): Invalid ID exceptions, empty user results
  - **State verification** (2 tests): Field preservation, isolated updates
  - **Ordering** (1 test): Descending timestamp sort validation

**Test Data (`NotificationsLocalRepositoryTestData.kt`):**

- Created reusable test fixtures with helper functions:
  - `createSampleNotification()`: Factory function with sensible defaults for all notification fields
  - `createTimestamp()`: Helper for timestamp creation, abstracting away Timestamp constructor details
  - 9 pre-configured notifications (3 per type: FRIEND_REQUEST, TODO_REMINDER, EVENT_REMINDER)
  - Distinct timestamps using incremental pattern (1000, 2000, 3000... seconds) for predictable ordering
  - Varied recipients (userA, userB, userC, userE) to test filtering logic
  - Organized into typed lists (`friendRequests`, `todoReminders`, `eventReminders`) for semantic test access

**Patterns:**

- Follows existing repository patterns (suspend functions, exception-based error handling)

## Testing?
All 12 tests pass. Achieved **100% line coverage** verified via coverage tool.

## Screenshots
image
100% Line coverage according to the coverage tool.

image
All tests pass.

## Anything Else?

**Next Steps:**
- Firebase repository implementation with Firestore integration
- Firebase repository tests (similar structure to local tests)
- **NotificationsViewModel with tests using this local repository**

### CODE REVIEW

File: app/src/main/java/com/android/gatherly/model/notification/NotificationsLocalRepository.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  override suspend fun deleteNotification(notificationId: String) {
+    if (notifications.remove(notificationId) == null) {
+      throw NoSuchElementException("Notification with ID $notificationId not found")
+    }
+  }
+
+  override suspend fun addNotification(notification: Notification) {
+    notifications[notification.id] = notification
```
Comment (claumnl) [inline app/src/main/java/com/android/gatherly/model/notification/NotificationsLocalRepository.kt]: NitPik: Here if the same id is reused by mistake for 2 notification, it silently overwrites the old one. This behavior is fine, but if you want to have more information on what could be going wrong in case of a bug, addNotification() could throw an exception if a notification already used this id. This is a minor non-blocking suggestion and not at all a requirement.

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  @Test
+  fun getUserNotifications_returnsNotificationsForSpecificUser() = runTest {
+    // Add multiple notifications for different users
+    repository.addNotification(friendRequests[0]) // userB
+    repository.addNotification(friendRequests[1]) // userB
+    repository.addNotification(todoReminders[0]) // userA
+    repository.addNotification(todoReminders[2]) // userC
+    repository.addNotification(eventReminders[0]) // userA
```
Comment (claumnl) [inline app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: NitPik: This snippet is used multiple times in the test file, extracting it in a helper could improve readability.

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    // Add notification, mark as read, verify wasRead is true
+    repository.addNotification(eventReminders[2])
+    repository.markAsRead(eventReminders[2].id)
+    val updatedNotification = repository.getNotification(eventReminders[2].id)
+    assertTrue(updatedNotification.wasRead)
+  }
+
+  @Test
+  fun deleteNotification_removesNotification() = runTest {
+    // TODO: Add notification, delete it, verify it's gone
```
Comment (claumnl) [inline app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: **NitPik:** This TODO comment in outdated, you implemented it, same in a couple of other tests.

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    repository.addNotification(todoReminders[0]) // time 4000
+    repository.addNotification(todoReminders[2]) // time 6000
+    repository.addNotification(eventReminders[0]) // time 8000
+
+    // Retrieve notifications for userA
+    val userANotifications = repository.getUserNotifications("userA")
+    assertEquals(2, userANotifications.size)
+
+    // Verify they are in descending order by emissionTime
+    assertTrue(userANotifications[0].emissionTime > userANotifications[1].emissionTime)
```
Comment (claumnl) [inline app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: Suggestion: You could use
 ```
assertEquals(
    userANotifications.sortedByDescending { it.emissionTime },
    userANotifications
)
```
To check the whole order and not just the order of the 2 first elements.

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt (hunk 2 above)
Comment (AACioffi) [reply app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: True - I'll probably change that.

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt (hunk 4 above)
Comment (AACioffi) [reply app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: I will change this for readability (even if the functionality seems identical)

File: app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    assertTrue(userENotifications.isEmpty())
+  }
+
+  // ============ State Verification ============
+
+  @Test
+  fun addNotification_preservesAllFields() = runTest {
+    // Add notification, retrieve it, verify all fields match
+    repository.addNotification(friendRequests[2])
+    assertEquals(friendRequests[2], repository.getNotification(friendRequests[2].id))
```
Comment (claumnl) [inline app/src/test/java/com/android/gatherly/model/notification/NotificationsLocalRepositoryTest.kt]: **Question:** Isn't this check redundant considering you check equality for each field in the lines below? If this is not intentional, either the general check or the per field checks can be safely removed.

**Review by claumnl — APPROVED** [review]
LGTM! Thank you for addressing my comments!
